llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,378 @@
+
+  // Get all the defined physical register that the MachineBasicBlock already
+  // uses. These physical register should not be used for replacement
+  // candidates, since this would introduce new WAW dependencies, which this
+  // pass removes.
+  SmallSet<MCPhysReg, 16> getDefinedPhysRegs(const MachineBasicBlock *MBB);
+
+  bool replaceReg(const Register Reg,
+                  SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs);
+
+  /// Find Free register of the same register Class type and
+  // exclude the forbidden Physical register from the result
+  SmallVector<MCPhysReg, 16>
+  getPriorityReplacementRegs(const Register Reg,
+                             const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs);
+
+  bool
+  isWorthRenaming(const Register &Reg, const MachineInstr &MI,
+                  const BitVector &UsedPhyRegs,
+                  const std::map<unsigned, SmallVector<const MachineInstr *, 4>>
+                      &SubRegDefs) const;
+
+  bool isIdentityCopy(const MachineInstr &MI);
+
+  bool isLastDef(const MachineOperand &Def, const MachineInstr &DefInstr,
+                 const MachineBasicBlock &MBB);
+
+  void addAliasRegs(SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                    const MCPhysReg &PhysReg);
+
+  Register getAssignedPhysReg(const MachineOperand &MO) const;
+};
+
+Register AIEWawRegRewriter::getAssignedPhysReg(const MachineOperand &MO) const {
+  Register PhysReg = MO.getReg();
+  if (PhysReg.isVirtual())
+    PhysReg = VRM->getPhys(PhysReg);
+  return PhysReg;
+}
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopMBBs =
+      AIELoopUtils::getInnermostLoopBlocks(MLI);
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+  Modified = false;
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopMBBs)
+    renameMBBPhysRegs(MBB);
+
+  return Modified;
+}
+
+bool hasSubRegCopy(
+    const unsigned ID,
+    const std::map<unsigned, SmallVector<const MachineInstr *, 4>>
+        &SubRegDefs) {
+  if (SubRegDefs.find(ID) == SubRegDefs.end())
+    return false;
+
+  for (const MachineInstr *MI : SubRegDefs.at(ID)) {
+    if (MI->isCopyLike())
+      return true;
+  }
+  return false;
+}
+
+std::map<unsigned, SmallVector<const MachineInstr *, 4>>
+getAllSubRegDefs(const MachineBasicBlock &MBB) {
+  std::map<unsigned, SmallVector<const MachineInstr *, 4>> SubRegDefs;
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand Def : MI.defs()) {
+      unsigned ID = Def.getReg().id();
```

Comment:
nit: the unsigned operator is is the same as id(), do you can remove .id(), the result is the same. I think you can also use `MCPhysReg` instead on `unsigned`.

Reply:
I should have made it clear, that here I just want virtual registers. Since this map is used to check, if any part of a Virtual Register was defined by a `copy `instruction. Changed the datatype and also added a check in the population of the map to make this distinction clear.